python/lsst/pipe/base/all_dimensions_quantum_graph_builder.py

```diff
@@ -542,3 +697,131 @@ def log_failure(self, log: LsstLogAdapter) -> None:
+        element name and then by the `DataCoordinate.required_values`
+        tuple.
+    """
+
+    def __init__(self, dimensions: DimensionGroup, dimension_records: DimensionRecordsMap):
+        self.dimensions = dimensions
+        self.indexers = {element: self._make_indexer(element) for element in dimensions.elements}
+        self.dimension_records = dimension_records
+
+    def _make_indexer(self, element: str) -> list[int]:
+        """Return a list of indexes into data ID full-values that extract the
+        `~DataCoordinate.required_values` for the ``element``
+        `DimensionElement.minimal_group` from the `~DataCoordinate.full_values`
+        for `dimensions`.
+        """
+        return [
+            self.dimensions._data_coordinate_indices[d]
+            for d in self.dimensions.universe[element].minimal_group.required
+        ]
+
+    def apply(self, data_id: DataCoordinate) -> DataCoordinate:
+        """Attach dimension records to the given data ID.
+
+        Parameters
+        ----------
+        data_id : `DataCoordinate`
+            Input data ID.
+
+        Returns
+        -------
+        expanded_data_id : `DataCoordinate`
+            Output data ID.
+        """
+        v = data_id.full_values
+        records_for_data_id = {}
+        for element, indexer in self.indexers.items():
+            records_for_element = self.dimension_records[element]
+            records_for_data_id[element] = records_for_element[tuple([v[i] for i in indexer])]
+        return data_id.expanded(records_for_data_id)
+
+    def maybe_apply(
+        self, data_id: DataCoordinate, leftovers: DataIdExpansionLeftovers
+    ) -> DataCoordinate | None:
+        """Attempt to attach dimension records to the given data ID, and record
+        the data IDs of missing dimension records on failure.
+
+        Parameters
+        ----------
+        data_id : `DataCoordinate`
+            Input data ID.
+        leftovers : `DataIdExpansionLeftovers`
+            Struct recording data IDs that were not expanded and the data IDs
+            of the dimension records that need to be fetched in order to do so.
+            The latter will be updated in place; callers are responsible for
+            updating the former when `None` is returned (since this method
+            doesn't have enough information to do so on its own).
+
+        Returns
+        -------
+        expanded_data_id : `DataCoordinate` or `None`
+            Output data ID, or `None` if one or more records were missing.
+        """
+        v = data_id.full_values
+        records_for_data_id = {}
+        failed = False
+        for element, indexer in self.indexers.items():
+            k = tuple([v[i] for i in indexer])
+            if (records_for_element := self.dimension_records.get(element)) is None:
+                leftovers.missing_record_data_ids[element].add(k)
+                failed = True
```

Comment:
Maybe add `break` or `return None` (and three lines below)?

Reply:
I think we can't shortcut early due to the need to fully populate `leftovers`.  I'll add a comment to that effect.